zebra-chain/src/addresses.rs

```diff
@@ -0,0 +1,222 @@
+//! Address types.
+
+use std::{fmt, io};
+
+use bs58;
+use ripemd160::{Digest, Ripemd160};
+use secp256k1::PublicKey;
+use sha2::Sha256;
+
+#[cfg(test)]
+use proptest_derive::Arbitrary;
+
+use crate::serialization::{SerializationError, ZcashDeserialize, ZcashSerialize};
+
+use crate::types::Script;
+
+/// A hash of a redeem script, as used in transparent
+/// pay-to-script-hash and pay-to-publickey-hash addresses.
+///
+/// The resulting hash in both of these cases is always exactly 20
+/// bytes. These are big-endian.
+/// https://en.bitcoin.it/Base58Check_encoding#Encoding_a_Bitcoin_address
+#[derive(Copy, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(Arbitrary))]
+pub struct AddressPayloadHash(pub [u8; 20]);
+
+impl<'a> From<&'a [u8]> for AddressPayloadHash {
+    fn from(bytes: &'a [u8]) -> Self {
+        let sha_hash = Sha256::digest(bytes);
+        let ripe_hash = Ripemd160::digest(&sha_hash);
+        let mut payload = [0u8; 20];
+        payload[..].copy_from_slice(&ripe_hash[..]);
+        Self(payload)
+    }
+}
+
+impl fmt::Debug for AddressPayloadHash {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("AddressPayloadHash")
+            .field(&hex::encode(&self.0))
+            .finish()
+    }
+}
+
+/// An enum describing the possible network choices.
+// XXX Stolen from zebra-network for now.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Network {
+    /// The production mainnet.
+    Mainnet,
+    /// The testnet.
+    Testnet,
+}
+
+/// Transparent Zcash Addresses
+///
+/// In Bitcoin a single byte is used for the version field identifying
+/// the address type. In Zcash two bytes are used. For addresses on
+/// the production network, this and the encoded length cause the first
+/// two characters of the Base58Check encoding to be xed as “t3” for
+/// P2SH addresses, and as “t1” for P2PKH addresses. (This does not
+/// imply that a transparent Zcash address can be parsed identically
+/// to a Bitcoin address just by removing the “t”.)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#transparentaddrencoding
+#[derive(Copy, Clone, Eq, PartialEq)]
+// #[cfg_attr(test, derive(Arbitrary))]
+pub enum TransparentAddress {
+    /// P2SH (Pay to Script Hash) addresses
+    PayToScriptHash {
+        /// Production, test, or other network
+        network: Network,
+        /// 20 bytes specifying a script hash.
+        script_hash: AddressPayloadHash,
+    },
+    /// P2PKH (Pay to Public Key Hash) addresses
+    PayToPublicKeyHash {
+        /// Production, test, or other network
+        network: Network,
+        /// 20 bytes specifying a public key hash, which is a RIPEMD-160
+        /// hash of a SHA-256 hash of a compressed ECDSA key encoding.
+        pub_key_hash: AddressPayloadHash,
```

Comment:
consistency: we should check whether elsewhere in the code uses `pub_key`/`PubKey` or `pubkey`/`Pubkey` and use one of them

Reply:
`secp256k1` defines `PublicKey`. `JoinSplit` has a `pub_key` field.